Kwabena/genx320 event mode

kwagyeman opened the pull request.

kwagyeman commented:
@nlyubova-psee - Can you take a look at this PR? I have some more testing to do with it. But, it's pretty much done. You should be able to run it on the H7 Plus or RT1062.

Please see the examples I added for how to use event mode. Note that generally, histogram mode (the default) produces much nicer looking images than event mode can do. I guess this is because the logic onboard the sensor is able to handle a much faster event data rate internally than what is possible with the EVT2.0 format over 8-bit parallel. So... the image generated by looking at the event datastream are somewhat of a subsampling of the true event rate.

I have some examples that the data rate without generating frames. For the H7 Plus, it can do about 250 FPS at max of 2048 event buffers. So, 512,000 events per second. This is less FPS than histogram mode, but, the processor has to due more work post-processing things.

Anyway, do you have any ideas for any kinda of simple things we can demonstrate on the event stream using numpy like array processing? I didn't have much inspiration beyond selecting all ON/OFF events.

nlyubova-psee commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_event_mode_filtering.py:
```diff
@@ -0,0 +1,65 @@
+# https://github.com/openmv/openmv/blob/master/LICENSE
+#
+# This example shows off using the genx320 event camera from Prophesee
+# using event streaming mode.
```
can we add something like this here : "and filtering events based on their polarity (ON/OFF)" ?  To make it more different from the previous example.

nlyubova-psee reviewed (commented):
Hi @kwagyeman , thank you for your work !
I can test it only on Monday (sorry for delay).

No worry about "good looking images" as they are useful only for people and not for algorithms. Good data are much more useful for algorithms rather than a beautiful image.

It is sad to hear that "it drops the next buffer coming in while it's busy working". Can we improve it somehow ? For example, if we tune biases to get less events, then processing should be theoretically faster. Will it still drop data in this case ?

We will think about a good example (maybe frequency measurement).

kwagyeman commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_event_mode_filtering.py:
```diff
@@ -0,0 +1,65 @@
+# to populate the ndarray. This takes some CPU time but is relatively
+# efficent. However, drawing the event histogram is VERY inefficient
+# on the CPU. FOR MAXIMUM PERFORMANCE YOU NEED TO PROCESS EVENTS USING
+# ULAB NDARRAY FUNCTIONS AND NOT DRAW THE EVENT HISTOGRAM.
```
Yeah, so, it's a random access to SDRAM where the image is stored. Each read/write accesses a completely different location. So, 2048 pixels being draw means up to 2048 new SDRAM row changes. These are costly because the entire architecture for the CPU is designed to access things in linear order.

Modern architectures can handle going backwards and forwards fine... albeit, forward access is guaranteed to be optimized in every silicon path and not just the CPU.

....

Given I'm already seeing performance issues... I'm thinking it will be faster to draw the image image using fb_alloc() in SRAM and then memcpy it to the final image location in SDRAM than trying to draw it in SDRAM directly.

nlyubova-psee replied:
In CDFrame, we can skip drawing events that happen for the same pixel over the accumulation period. (in this case you need to draw less than 2048 pixels)
In case of large accumulation periods (that is the case for 250fps), usually more than 1 event happens per pixel and we do not need to draw all of them. 

If you have already considered it, then please ignore my comment.

kwagyeman replied:
From a performance standpoint, there's no difference in incrementing a counter per pixel position or just setting it once. So, might as well go with a counter versus a flag as this gives you 8-bit values versus just 0/255. Storing the histogram as a binary value (1-bit) would be slower due to the shifting an masking per pixel.

kwagyeman commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_event_mode.py:
```diff
@@ -0,0 +1,59 @@
+    # added to them for PIX_ON_EVENT events and subtracted from them for
+    # PIX_OFF_EVENT events clampped between 0 and 255. Pass clear=False to keep
+    # accumulating events in the histogram image.
+    img.draw_event_histogram(events, clear=True, brightness=128, contrast=64)
```
Renamed

iabdalkader commented on modules/py_image.c:
```diff
@@ -1269,7 +1269,7 @@ static MP_DEFINE_CONST_FUN_OBJ_KW(py_image_save_obj, 2, py_image_save);
 static mp_obj_t py_image_flush(mp_obj_t img_obj) {
     framebuffer_t *fb = framebuffer_get(0);
-    framebuffer_update_jpeg_buffer(fb);
+    framebuffer_update_jpeg_buffer(fb, py_image_cobj(img_obj));
```
It was meant to update the FB after drawing to it, since all images originally were FB pointers. I think the cleaner way to fix this would be to refactor `framebuffer_update_jpeg_buffer` to accept an image, since that's what it really expects. You could probably reuse the image pointer passed to every `snapshot`.

kwagyeman replied:
I was actually going to do that initially, but, it calls a bunch of functions internally that need the FB pointer. However, none of the functions it passes the FB pointer to actually use the pointer... so, I can drop the pointer from those functions and almost everything in framebuffer_update_jpeg_buffer. 

But, it does check `fb->streaming_enabled` internally. How should I handle this? This expands the scope of the refactor.

kwagyeman replied:
I thought about how to refactor this to get rid of the fb_streaming_enabled issue. Will send this in another PR.

iabdalkader replied:
Let's see it, but on second look it does expect a framebuffer, so maybe just copy the image to be flushed to fb if it's not already the fb ? And let's remove the `src_image` confusing stuff, it's not needed.

kwagyeman replied:
> so maybe just copy the image to be flushed to fb if it's not already the FB

Not possible, the FB in event mode is like 1024x4 pixels. Have to decouple.

iabdalkader replied:
Not following? When calling `img.flush()` if the image != FB you could just memcpy it to fb, before calling `update_jpeg_buffer`.

kwagyeman replied:
In event mode the fb is like 1024x4 bytes pixels which represents 1024 events (4-bytes per event). The generated image is 320x320 pixels. So, I can't even put the image in the frame buffer. I need to directly flush the image into the jpeg buffer.

iabdalkader replied:
I see the problem. Okay, we could refactor `framebuffer_update_jpeg_buffer` to call a new function that accepts `image_t`, which could also be called by flush. `framebuffer_update_jpeg_buffer` can still do all the checks, like `streaming_enabled` etc.. before calling the new function. That mode that prints the image to serial, how useful is it? We could also remove it.

iabdalkader replied:
> I'm thinking dropping it and just printing a message that the frame buffer can't be displayed because the image is too large

Yes we should drop it. As for the message, and text dropping, I don't like printing a lot of messages. Can we draw the error on the frame buffer instead?

> As for the refactor, I was actually going to completely remove the streaming_enabled and the logic in the omv module that controls it as you can just pass update=False to sensor now. So, it's not needed at all anymore.

Okay, I never really liked it, seems confusing to have to flags that control streaming.

kwagyeman replied:
> Yes we should drop it. As for the message, and text dropping, I don't like printing a lot of messages. Can we draw the error on the frame buffer instead?

Good idea. I can send an escape sequence and have OpenMV IDE decode it.

kwagyeman replied:
https://github.com/openmv/openmv/pull/2755 - Will rebase after PR is merged.

kwagyeman commented:
Only the ndarray manipulation needs a collection. Otherwise, everything I've created for allocating and writing to the ndarray doesn't. Not much you can do about memory allocations if when using ULAB.

kwagyeman commented:
Okay, I've resolved the old events issue. Just needed to slice the output event array by the valid event count which is a very fast operation.

https://github.com/user-attachments/assets/b8ba5550-e05e-4c64-be3a-89463acf35a6

Just need to check that everything is working now on the RT1062 and N6 and this it's ready for review.

nlyubova-psee commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_event_mode_deep_buffer.py:
```diff
@@ -0,0 +1,55 @@
+# https://github.com/openmv/openmv/blob/master/LICENSE
+#
+# This example shows off using the genx320 event camera from Prophesee
+# using event streaming mode.
```
Could you please update the description to make the sample more clear (for example say that the buffer size is very large)?

nlyubova-psee reviewed (approved):
Even if we plan to test the data integrity, the PR is general is ok for me.

kwagyeman commented:
@nlyubova-psee - I've updated the examples with your requested changes. I've also gone ahead and removed support for this new feature in the old sensor API which is now deprecated in favor of the new CSI module and updated all the examples to match this. I've retested them all and they work.

Please let me know what you mean by "timesurface", like, if you just max out the contrast at 256 then you'll get an image where events are only the latest value as it will swing the histogram between 0 and 255 (adding +256 or -256). However, this doesn't seem like it needs a separate example.

Also, should we disable the AFK by default? It makes the image blocky often.

Anyway, please let me know any other thoughts. I think the PR is just about ready.

nlyubova-psee commented:
> Also, should we disable the AFK by default? It makes the image blocky often.

Agree I prefer to deactivate it. 

> Please let me know what you mean by "timesurface", like, if you just max out the contrast at 256 then you'll get an image where events are only the latest value as it will swing the histogram between 0 and 255 (adding +256 or -256). However, this doesn't seem like it needs a separate example.

Timasurface is a 2D image where each pixel has the timestamp of the last event happened for this pixel. Thus, it integrates some history of motion over the frame. Visually it maybe does not help :) but algorithms can profit of this time information. Anyway, it is not important to do it. Just an idea.

kwagyeman commented:
Okay, other than disabling AFK by default. Please test the examples and let me know if the PR is ready for Ibrahim to finish his code review.

kwagyeman commented:
@nlyubova-psee - Can you finish your review of this PR this week? Customers are asking...

iabdalkader commented:
This pull request has been mentioned on **OpenMV Forums**. There might be relevant details there:

https://forums.openmv.io/t/how-to-achieve-high-speed-led-frequency-detection-10s-of-khz-with-genx320-on-openmv-h7-plus/11207/11

iabdalkader commented on drivers/sensors/genx320.c:
```diff
@@ -581,6 +589,106 @@ static int ioctl(omv_csi_t *csi, int request, va_list ap) {
+                break;
+            }
+
+            if (is_valid_framesize(csi->framesize, true) < 0) {
```
If I'm following this correctly, you should probably check here that the frame size is >= the ndarray size.

kwagyeman replied:
The py_csi_ng layer checks that the array matches the ndarray size. So, you get a nice error message about the array size being wrong.

> Note that the ioctl could just set the frame size internally, based on the ndarray size, which would make it impossible for the user to set one res and pass a smaller ndarray. Though I'm torn between that and staying consistent with all the other scripts.

Yeah, it's possible, but, would need to make sure the ndarray passed a list of qualifications in the py_csi_ng layer. Like, I have to keep the dma sizes, line buffers, and etc. in mind for the frame sizes. Not impossible to handle though, but, it kinda means you aren't passing a frame size and pixformat anymore which makes things inconsistent with other scripts.

iabdalkader replied:
Ah I see it's validate in csi.

> makes things inconsistent with other scripts as you noticed.
Yeah, let's keep it this way.

iabdalkader commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_event_mode.py:
```diff
@@ -0,0 +1,56 @@
+    # Draws the events on the image. If clear=True then the image is cleared
+    # to the default value of "brightness". Event histogram bins have "contrast"
+    # added to them for PIX_ON_EVENT events and subtracted from them for
+    # PIX_OFF_EVENT events clampped between 0 and 255. Pass clear=False to keep
```
```suggestion
    # PIX_OFF_EVENT events clamped between 0 and 255. Pass clear=False to keep
```

iabdalkader commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_event_mode.py:
```diff
@@ -0,0 +1,56 @@
+events = np.zeros((2048, 6), dtype=np.uint16)
+
+# Initialize the sensor.
+csi0 = csi.CSI(cid=csi.GENX320)
```
```suggestion
csi0 = csi.CSI()
```
You only have to specify the ID if it's used with another sensor, otherwise you get the default and only sensor.

kwagyeman replied:
Yes, my thought here is that by specifying the cid, the script will work for the upcoming dual module too. Is that sensible?

iabdalkader replied:
Yes, good point. Though it will probably also require its own example(s).

kwagyeman commented:
Prophesee informed me they will have completed a final review of things by mid-next week.

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -82,50 +76,69 @@
-    psee_sensor_write(csi, CPI_PACKET_TIME_CONTROL, ACTIVE_SENSOR_WIDTH << CPI_PACKET_TIME_CONTROL_PERIOD_Pos |
-                      HSYNC_CLOCK_CYCLES << CPI_PACKET_TIME_CONTROL_BLANKING_Pos);
-    psee_sensor_write(csi, CPI_FRAME_SIZE_CONTROL, ACTIVE_SENSOR_HEIGHT);
+    int packet_width = enable ? resolution[csi->framesize][0] : ACTIVE_SENSOR_WIDTH;
+    int packet_height = enable ? resolution[csi->framesize][1] : ACTIVE_SENSOR_HEIGHT;
+    int packet_hsync = enable ? EVENT_HSYNC_CLOCK_CYCLES : HISTO_HSYNC_CLOCK_CYCLES;
```
Done

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -159,40 +173,57 @@ static int reset(omv_csi_t *csi) {
     return 0;
 }
 
+static int reset(omv_csi_t *csi) {
+    csi->color_palette = NULL;
+
+    reset_called = false;
+    event_mode = false;
+    contrast = CONTRAST_DEFAULT;
+    brightness = BRIGHTNESS_DEFAULT;
+    event_time = 0;
+
+    // Set histogram mode by default.
+    return set_event_mode(csi, false);
+}
+
```
Done

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -159,40 +173,57 @@ static int reset(omv_csi_t *csi) {
-    #else
-    psee_sensor_start(csi, &dcmi_evt);
-    #endif //  (OMV_GENX320_EHC_ENABLE == 1)
+
+    reset_called = true;
+    event_mode = enable;
```
Done

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -590,10 +698,17 @@ static int ioctl(omv_csi_t *csi, int request, va_list ap) {
     // Initialize csi structure
     csi->reset = reset;
     csi->sleep = sleep;
+    csi->match = match;
```
Done

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -82,50 +76,69 @@
-#if (OMV_GENX320_CAL_ENABLE == 1)
-static bool hot_pixels_disabled = false;
-#endif // (OMV_GENX320_CAL_ENABLE == 1)
+static bool reset_called = false;
+static bool event_mode = false;
 static int32_t contrast = CONTRAST_DEFAULT;
 static int32_t brightness = BRIGHTNESS_DEFAULT;
+static uint64_t event_time = 0;
```
Done

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -82,50 +76,69 @@
+    return -1;
+}
+
+static int set_event_mode(omv_csi_t *csi, bool enable) {
+    if (reset_called) {
+        if (event_mode == enable) {
+            return 0;
+        }
+
+        psee_sensor_stop(csi, event_mode ? &dcmi_evt : &dcmi_histo);
+        psee_sensor_destroy(csi);
+    }
```
Done, the mode check has to be gated though by the the issd being valid otherwise you'd do nothing for the first call.

kwagyeman replied:
Now updated.

kwagyeman commented:
Okay, I fixed the bugs with the PR. It's now working on the H7 Plus, RT1062, and N6 correctly. Event mode stopped working because with the new snapshot code, omv_csi_snapshot() has to be called versus the snapshot platform implementation; otherwise, the buffer management doesn't happen.

The PR is now rebased on master too.

iabdalkader commented:
> Let me know if you have a better idea for this.

I don't follow. That flag is only set in event mode, so check if it's in event mode?

kwagyeman commented:
Yeah, I think preventing alternation fixes the issue. Setting the frame size to one of the events modes is enough for me to know if we are in event or histogram mode.

iabdalkader commented:
> Setting the frame size to one of the events modes is enough for me to know if we are in event or histogram mode.

Do we need to set pixformat/framesize at all for event mode if snapshot will not be called? Can't you infer the size of the buffer from read_events?

iabdalkader commented:
> It's more than the whole camera driver system needs the frame size set. Lots of code configures the DMA system based on the frame size

Set it from the driver? Just like what you have been doing for snapshot? I feel an ioctl to set mode + read events should would make a very clear and concise API:

```Python
csi0.reset()
csi0.ioctl(SET_MODE, EVENT)? 

while True:
   csi0.ioctl(READ_EVENTS, ...)
```

kwagyeman commented:
Still working on this, will have it done tomorrow.

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -160,39 +189,61 @@ static int reset(omv_csi_t *csi) {
+    genx->brightness = BRIGHTNESS_DEFAULT;
+    genx->event_time_us = 0;
+    genx->issd = NULL;
+    genx->mode = GENX_MODE_INVALID;
```
Okay, resolved this, with set_mode, everything became simpler, and I was able to remove resetting these variables.

kwagyeman commented:
Set mode is implemented, code is simpler too. API is as suggested now.

iabdalkader commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_event_mode_calibration.py:
```diff
@@ -0,0 +1,68 @@
+    # added to them for PIX_ON_EVENT events and subtracted from them for
+    # PIX_OFF_EVENT events clamped between 0 and 255. Pass clear=False to keep
+    # accumulating events in the histogram image.
+    img.draw_event_histogram(events[:event_count], clear=True, brightness=128, contrast=64)
```
Should we call this just `draw_events`? And have it flush the buffer?

iabdalkader replied:
> I didn't want to call it draw_events() as I think that would be like drawing crosses and etc.

The events are not a generic histogram right? If they were we would have called it `draw_histogram`. The function draws something very specific with its format. How about `draw_ec_events` ? Or `draw_genx_events`?

kwagyeman replied:
Maybe `draw_ec_event_histogram`? I made the ec event stuff to abstract away from the genx320 so everything outside of the camera driver has a more generic name to allow us to support future higher resolution event cameras.

Fine with `draw_ec_events` too, drawing events in a histogram is the only way to really visualize them, the histogram part is redundant.

iabdalkader replied:
Yeah `draw_ec_events` is fine. What does EC stand for?

iabdalkader replied:
Actually, I think it should be `draw_histogram` you don't pass the whole events array right? Just a slice which is probably the histogram?

iabdalkader replied:
Let's just leave it as is, resolving this.

iabdalkader commented on drivers/sensors/genx320.c:
```diff
@@ -563,24 +580,152 @@ static int ioctl(omv_csi_t *csi, int request, va_list ap) {
+                break;
+            }
+
+            // Invalidate frame.
+            csi->fb->pixfmt = PIXFORMAT_INVALID;
+
+            ret = event_post_process(genx, (ec_event_t *) va_arg(ap, ec_event_t *), &image);
```
I see, it's also called from `OMV_CSI_IOCTL_GENX320_CALIBRATE`. It's still possible though, if you like, you could just store the ndarray pointer in `priv`. If set do post_process. And calibrate should set it to NULL.

kwagyeman commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_event_mode_calibration.py:
```diff
@@ -0,0 +1,68 @@
+# Surface to draw the histogram image on.
+img = image.Image(320, 320, image.GRAYSCALE)
+
+# ndarray to hold events from the camera
+# must be EVT_res events by 6 values
+#
+# 0: event type
+# 1: seconds timestamp
+# 2: milliseconds timestamp
+# 3: microseconds timestamp
+# 4: x coordinate (0-319 for the genx320)
+# 5: y coordinate (0-319 for the genx320)
+events = np.zeros((2048, 6), dtype=np.uint16)
```
Done

kwagyeman commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_event_mode_calibration.py:
```diff
@@ -0,0 +1,68 @@
+# Initialize the sensor.
+csi0 = csi.CSI(cid=csi.GENX320)
+csi0.reset()
+csi0.ioctl(csi.IOCTL_GENX320_SET_MODE, csi.GENX320_MODE_EVENT,
+           csi.EVT_2048)  # Must be EVT_1024/2048/.../65536
```
Done

kwagyeman commented on common/omv_csi.h:
```diff
@@ -188,6 +188,16 @@ typedef enum {
     OMV_CSI_FRAMESIZE_WQXGA,    // 2560x1600
     OMV_CSI_FRAMESIZE_WQXGA2,   // 2592x1944
+    #if (OMV_GENX320_ENABLE == 1)
+    // Event Resolutions
```
Done

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -82,50 +76,73 @@
 
-static int reset(omv_csi_t *csi) {
-    csi->color_palette = NULL;
+static int genx_set_mode(omv_csi_t *csi, genx_mode_t mode, int framesize) {
```
Done

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -82,50 +76,73 @@
-    psee_sensor_write(csi, CPI_FRAME_SIZE_CONTROL, ACTIVE_SENSOR_HEIGHT);
+    uint32_t packet_width = resolution[framesize][0];
+    uint32_t packet_height = resolution[framesize][1];
+    uint32_t packet_hsync = (mode == OMV_CSI_GENX320_MODE_EVENT) ? EVENT_HSYNC_CLOCK_CYCLES : HISTO_HSYNC_CLOCK_CYCLES;
+    psee_sensor_write(csi, CPI_PACKET_SIZE_CONTROL, packet_width);
+    psee_sensor_write(csi, CPI_PACKET_TIME_CONTROL, packet_width << CPI_PACKET_TIME_CONTROL_PERIOD_Pos |
+                      packet_hsync << CPI_PACKET_TIME_CONTROL_BLANKING_Pos);
```
Done

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -341,127 +394,91 @@ static void disable_hot_pixels(omv_csi_t *csi, uint8_t *histogram) {
+    return valid_count;
+}
+
+static int snapshot(omv_csi_t *csi, image_t *image, uint32_t flags) {
```
Done

kwagyeman commented:
Okay, all changes done. Custom resolution mode is working great.

iabdalkader commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_histogram_mode_grayscale_led_tracking.py:
```diff
@@ -9,24 +9,25 @@
+csi0 = csi.CSI(cid=csi.GENX320)
+csi0.reset()
+csi0.pixformat(csi.GRAYSCALE)  # Must always be grayscale.
+csi0.framesize(csi.B320X320)  # Must always be 320x320.
```
I removed all these custom resolution constants in https://github.com/openmv/openmv/pull/2804 let's merge that first and then just use the w/h directly.

```suggestion
csi0.framesize((320, 320))
```

kwagyeman replied:
Done, testing everything now.

kwagyeman commented on scripts/examples/01-Camera/03-Event-Cameras/02-Genx320/genx320_event_mode_deep_buffer_filtering.py:
```diff
@@ -0,0 +1,63 @@
+# 2: milliseconds timestamp
+# 3: microseconds timestamp
+# 4: x coordinate (0-319 for the genx320)
+# 5: y coordinate (0-319 for the genx320)
+events = np.zeros((32768, 6), dtype=np.uint16)  # Must be 1024/2048/.../65536
```
Done

kwagyeman commented on drivers/sensors/genx320.c:
```diff
@@ -590,10 +704,32 @@ static int ioctl(omv_csi_t *csi, int request, va_list ap) {
     return ret;
 }
 
+static int snapshot(omv_csi_t *csi, image_t *image, uint32_t flags) {
```
Done
